### PR TITLE
Scripts/EasternKingdoms Dun Morogh all quests working, all but one 10…

### PR DESCRIPTION
**Pull request by AussieGlenn**
…0% blizzlike, all vendors working, major contributions by @Cayden

**Changes proposed:**
Fix every quest and vendor in Dun Morogh
Remove invalid quest mailouts for Riding Pamphlets
**Issues addressed:** Closes #  (insert issue tracker number)

**Tests performed:** (Does it build, tested in-game, etc.)
Compiled and full test run with human mage and dwarf rogue

**Known issues and TODO list:** (add/remove lines as needed)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by AussieGlenn** [comment]
Hi @Traesh ,

Please make changes as you see fit, and credit @Cayden as a major contributor.

Thanks,

Glenn

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 1]
```diff
@@ -18,6 +18,19 @@
...
+#include "SpellScript.h"
+#include "Player.h"
+#include "Vehicle.h"
+#include "ObjectMgr.h"
+#include "Unit.h"
+#include "TemporarySummon.h"
+#include "GameObject.h"
+#include "CombatAI.h"
+#include "ObjectAccessor.h"
+#include "ScriptedEscortAI.h"
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp:35]: so lot includes? rly? :D

File: src/server/scripts/EasternKingdoms/eastern_kingdoms_script_loader.cpp [hunk 2]
```diff
@@ -477,6 +479,7 @@ void AddEasternKingdomsScripts()
     AddSC_zone_gnomeregan();
     AddSC_ghostlands();
     AddSC_hinterlands();
+	AddSC_ironforge();
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/eastern_kingdoms_script_loader.cpp]: tabs to spaces, same below

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 3]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+
+	class spell_burn_constriction_totem_SpellScript : public SpellScript
+	{
+		PrepareSpellScript(spell_burn_constriction_totem_SpellScript);
+
+		void HandleScript(SpellEffIndex /*effIndex*/)
+		{
+			if (!GetHitUnit())
+				return;
+			GetCaster()->ToPlayer()->RewardPlayerAndGroupAtEvent(NPC_CONSTRICTION_TOTEM, GetCaster());
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: GetCaster()->ToPlayer() possible crash if no check if caster is player

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 4]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+			GetHitUnit()->ToCreature()->DisappearAndDie();
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: same, possible crash

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 5]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+
+    class spell_extinguish_fire_SpellScript : public SpellScript
+    {
+        PrepareSpellScript(spell_extinguish_fire_SpellScript);
+
+        void HandleScript(SpellEffIndex /*effIndex*/)
+        {
+            if (!GetHitUnit())
+                return;
+            GetCaster()->ToPlayer()->RewardPlayerAndGroupAtEvent(NPC_BURNING_BOMBER_CREDIT, GetCaster());
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: same possible crashes as abowe

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 6]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+        PrepareSpellScript(spell_emote_stolen_ram_SpellScript);
+
+        void HandleDummy(SpellEffIndex /*effIndex*/)
+        {
+            Unit* hitUnit = GetHitUnit();
+            if (!hitUnit)
+                return;
+
+            if (hitUnit->GetTypeId() != TYPEID_PLAYER)
+                return;
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: if (hitUnit->GetTypeId() != TYPEID_PLAYER)  ->   if (!hitUnit->IsPlayer())

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 7]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+
+            if (hitUnit->ToPlayer()->GetQuestStatus(QUEST_RAMS_ON_THE_LAM) != QUEST_STATUS_INCOMPLETE)
+                return;
+
+            GetCaster()->GetMotionMaster()->MovePoint(0, hitUnit->GetPosition());
+
+            if (GetCaster()->ToCreature())
+                GetCaster()->ToCreature()->DespawnOrUnsummon(3000);
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: if (auto const& creature = GetCaster()->ToCreature())
                creature ->DespawnOrUnsummon(3000);

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 8]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+        {
+            if (player->GetDistance(me) > 15.f)
+                return;
+
+            switch (emoteId)
+            {
+            case TEXT_EMOTE_WHISTLE:
+            case TEXT_EMOTE_WAVE:
+                DoCast(player, SPELL_ON_WAVEEMOTE_TO_STOLEN_RAM);
+                break;
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: any default: break; ?

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 9]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+    {
+        npc_sergeant_flinthammerAI(Creature* creature) : ScriptedAI(creature)
+        {
+        }
+
+        ObjectGuid Troll1GUID;
+        ObjectGuid Troll2GUID;
+        ObjectGuid Troll3GUID;
+        ObjectGuid Troll4GUID;
+        ObjectGuid PlayerGUID;
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: ObjectGuid Troll1GUID;
        ObjectGuid Troll2GUID;
        ObjectGuid Troll3GUID;
        ObjectGuid Troll4GUID;

make array? 
std::array<ObjectGuid, 4> or ObjectGuid [4]

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 10]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+                        Phase++;
+                    }
+
+                    case 8:
+                    {
+                        Creature* Troll1 = ObjectAccessor::GetCreature(*me, Troll1GUID);
+                        Creature* Troll2 = ObjectAccessor::GetCreature(*me, Troll2GUID);
+                        Creature* Troll3 = ObjectAccessor::GetCreature(*me, Troll3GUID);
+                        Creature* Troll4 = ObjectAccessor::GetCreature(*me, Troll4GUID);
+                        Troll1->SetReactState(REACT_AGGRESSIVE);
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: im sure that it'll be easyer with array of Trolls 🗡 
for (auto const& trolGuid : guidsArray)
{
    if (auto const& troll = = ObjectAccessor::GetCreature(*me, trolGuid ))
{
troll ->SetReactState(REACT_AGGRESSIVE);
troll ->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC);
}

}

File: src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp [hunk 11]
```diff
@@ -83,7 +96,639 @@ class npc_frozen_mountaineer : public CreatureScript
...
+        }
+
+    private:
+    };
+
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return new npc_sergeant_flinthammerAI(creature);
+    }
+};
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_dun_morogh.cpp]: use mby RegistSpellScrept / RegisterNpcScript ? it'll make just 300 lines of code insted of 500

File: src/server/scripts/EasternKingdoms/zone_ironforge.cpp [hunk 12]
```diff
@@ -16,10 +16,176 @@
...
+#include "SpellScript.h"
 #include "Player.h"
+#include "Vehicle.h"
+#include "ObjectMgr.h"
+#include "Unit.h"
+#include "TemporarySummon.h"
+#include "GameObject.h"
+#include "CombatAI.h"
+#include "ObjectAccessor.h"
+#include "ScriptedEscortAI.h"
```
Comment (soulfrost) [inline src/server/scripts/EasternKingdoms/zone_ironforge.cpp]: too lot of uselles and unneeded includes

**Comment by AussieGlenn** [comment]
Hi @soulfrost ,

Thanks for the suggestions, my c++ is crap, I just copy what is working elsewhere and paste.

It works, and will need to be improved upon before committing to the core.

Thanks,

Glenn

**Comment by vipwow1990** [comment]
This spell causes the server to crash
http://www.wowhead.com/spell=207684/sigil-of-musery
How can this be remedied?

**Comment by AussieGlenn** [comment]
Hi @vipwow1990,

Issues can be reported via our Mantis issue tracker at https://bt.ashamane.com

**Comment by vipwow1990** [comment]
@AussieGlenn Thank you!

**Comment by boom8866** [comment]
@soulfrost  If you see my information, please contact me. I have something urgent to ask you

**Pull request merged**